src/tunnel.cc

```diff
@@ -1285,21 +1300,32 @@ TunnelStateData::noteDestinationsEnd(ErrorState *selectionError)
     // if all of them fail, tunneling as whole will fail
     Must(!selectionError); // finding at least one path means selection succeeded
 
-    if (usingDestination()) {
-        // We are already using a previously opened connection but also
-        // receiving destinations in case we need to re-forward.
-        Must(!transportWait);
+    if (transportWait) {
+        assert(!transporting());
+        notifyConnOpener();
+        return; // and continue to wait for the noteConnection() callback
+    }
+
+    if (transporting()) {
```

Comment:
The assert here makes me wonder if we should reverse the order of these if-statements. I do not insist on any change here.

Reply:
Ah. IMO exclusivity is better represented by if-else structure. This sequential style can look nice, but does imply some order relations to the statements - which is prone to devs "improving" the sequence for performance goals.

Anyway, closing the thread.

Reply:
> Ah. IMO exclusivity is better represented by if-else structure.

The `if-else` structure is indeed the best for representing an exclusive `c` and `not c` relationship (when handling of each condition requires very little code). Unfortunately, this code has nothing to do with that kind of relationship.

Here, instead of two mutually exclusive conditions, we have three: `c1`, `c2`, and `neither c1 nor c2`. C++ does not have syntax for expressing such mutual exclusivity well. And no, `if-else-if-else` does not express such mutual exclusivity (at all).

> This sequential style can look nice, but does imply some order relations to the statements - which is prone to devs "improving" the sequence for performance goals.

I do not think your assumption is correct. The clauses in _both_ if-else-if-else and if-return-if-return styles may imply some order relation _and_ may be reordered for performance (or other) reasons.

C++ just does not give us a good way to express mutual exclusivity of three conditions. The current code is the best we can do with the tools we have AFAICT.